File: src/shared/BasicCommerce.Domain/Entities/PurchaseOrder.cs
using BasicCommerce.Domain.Enums;
using BasicCommerce.Domain.Exceptions;

namespace BasicCommerce.Domain.Entities;

public class PurchaseOrder : TenantEntity
{
    public string OrderNumber { get; private set; } = default!;
    public Guid SupplierId { get; private set; }
    public Guid WarehouseId { get; private set; }
    public PurchaseOrderStatus PurchaseOrderStatus { get; private set; } = PurchaseOrderStatus.Draft;
    public DateTime OrderDate { get; private set; }
    public DateTime? ExpectedDate { get; private set; }
    public DateTime? ReceivedDate { get; private set; }
    public string? Notes { get; private set; }
    public string Currency { get; private set; } = "BDT";

    public Supplier? Supplier { get; private set; }
    public Warehouse? Warehouse { get; private set; }

    private readonly List<PurchaseOrderItem> _items = [];
    public IReadOnlyCollection<PurchaseOrderItem> Items => _items.AsReadOnly();

    public decimal TotalAmount => _items.Sum(i => i.TotalCost);

    private PurchaseOrder() { }

    public static PurchaseOrder Create(Guid tenantId, string orderNumber, Guid supplierId,
        Guid warehouseId, DateTime orderDate, DateTime? expectedDate = null,
        string? notes = null, string currency = "BDT")
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(orderNumber);
        return new PurchaseOrder
        {
            TenantId = tenantId,
            OrderNumber = orderNumber,
            SupplierId = supplierId,
            WarehouseId = warehouseId,
            OrderDate = orderDate,
            ExpectedDate = expectedDate,
            Notes = notes,
            Currency = currency
        };
    }

    public void AddItem(Guid productId, decimal orderedQuantity, decimal unitCost)
    {
        if (PurchaseOrderStatus != PurchaseOrderStatus.Draft)
            throw new DomainException("Items can only be added to a Draft purchase order.");
        var existing = _items.FirstOrDefault(i => i.ProductId == productId);
        if (existing is not null)
            existing.UpdateQuantity(orderedQuantity, unitCost);
        else
            _items.Add(PurchaseOrderItem.Create(Id, productId, orderedQuantity, unitCost));
    }

    public void RemoveItem(Guid productId)
    {
        if (PurchaseOrderStatus != PurchaseOrderStatus.Draft)
            throw new DomainException("Items can only be removed from a Draft purchase order.");
        var item = _items.FirstOrDefault(i => i.ProductId == productId)
            ?? throw new NotFoundException("PurchaseOrderItem", productId);
        _items.Remove(item);
    }

    public void Submit()
    {
        if (PurchaseOrderStatus != PurchaseOrderStatus.Draft)
            throw new DomainException("Only Draft orders can be submitted.");
        if (_items.Count == 0)
            throw new DomainException("Cannot submit a purchase order with no items.");
        PurchaseOrderStatus = PurchaseOrderStatus.Submitted;
        UpdatedAt = DateTime.UtcNow;
    }

    public void MarkPartiallyReceived()
    {
        if (PurchaseOrderStatus is not (PurchaseOrderStatus.Submitted or PurchaseOrderStatus.PartiallyReceived))
            throw new DomainException("Only Submitted or PartiallyReceived orders can be partially received.");
        PurchaseOrderStatus = PurchaseOrderStatus.PartiallyReceived;
        UpdatedAt = DateTime.UtcNow;
    }

    public void MarkReceived()
    {
        PurchaseOrderStatus = PurchaseOrderStatus.Received;
        ReceivedDate = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Cancel()
    {
        if (PurchaseOrderStatus is PurchaseOrderStatus.Received)
            throw new DomainException("Received orders cannot be cancelled.");
        PurchaseOrderStatus = PurchaseOrderStatus.Cancelled;
        UpdatedAt = DateTime.UtcNow;
    }

    public void UpdateNotes(string? notes)
    {
        Notes = notes;
        UpdatedAt = DateTime.UtcNow;
    }
}

File: src/clients/BasicCommerce.WebAdmin/src/types/purchaseOrder.ts
export type PurchaseOrderStatus =
  | 'Draft'
  | 'Submitted'
  | 'PartiallyReceived'
  | 'Received'
  | 'Cancelled';

export interface PurchaseOrderItem {
  id: string;
  productId: string;
  productName: string;
  sku: string;
  isProductActive: boolean;
  orderedQuantity: number;
  receivedQuantity: number;
  remainingQuantity: number;
  unitCost: number;
  totalCost: number;
  isFullyReceived: boolean;
}

export interface PurchaseOrderSummary {
  id: string;
  orderNumber: string;
  supplierId: string;
  supplierName: string;
  warehouseId: string;
  warehouseName: string;
  status: PurchaseOrderStatus;
  orderDate: string;
  expectedDate: string | null;
  currency: string;
  totalAmount: number;
  itemsCount: number;
}

export interface PurchaseOrderDetail extends PurchaseOrderSummary {
  receivedDate: string | null;
  notes: string | null;
  items: PurchaseOrderItem[];
}

export interface PurchaseOrderListResponse {
  items: PurchaseOrderSummary[];
  totalCount: number;
  page: number;
  pageSize: number;
}

export interface CreatePOLineItem {
  productId: string;
  productName: string;
  sku: string;
  isProductActive?: boolean;
  quantity: number;
  unitCost: number;
}

export interface CreatePurchaseOrderFormData {
  supplierId: string;
  warehouseId: string;
  orderDate: string;
  expectedDate: string;
  notes: string;
  currency: string;
  items: CreatePOLineItem[];
}

export interface ReceiveLineItem {
  productId: string;
  receivedQuantity: number;
}

export const PO_STATUS_CONFIG: Record<
  PurchaseOrderStatus,
  { label: string; color: string; dot: string }
> = {
  Draft:             { label: 'Draft',              color: 'bg-gray-100 dark:bg-gray-800 text-gray-600 dark:text-gray-400',               dot: 'bg-gray-400' },
  Submitted:         { label: 'Submitted',           color: 'bg-blue-100 dark:bg-blue-900/30 text-blue-700 dark:text-blue-300',             dot: 'bg-blue-500' },
  PartiallyReceived: { label: 'Partially Received',  color: 'bg-amber-100 dark:bg-amber-900/30 text-amber-700 dark:text-amber-300',         dot: 'bg-amber-500' },
  Received:          { label: 'Received',            color: 'bg-emerald-100 dark:bg-emerald-900/30 text-emerald-700 dark:text-emerald-300', dot: 'bg-emerald-500' },
  Cancelled:         { label: 'Cancelled',           color: 'bg-red-100 dark:bg-red-900/30 text-red-600 dark:text-red-400',                dot: 'bg-red-400' },
};
